### PR TITLE
#908 - Fix Safari TP and @webcomponents/custom-element polyfill

### LINKED ISSUE #908: Fix Safari TP
This is because of a bug in document-register-element where it's always polyfilling and doesn't seem to find the constructor properly to initialise. I didn't have time to investigate it, so I switched to using the webcomponents polyfill instead, which got us past that but now there's two more errors:

- The `@webcomponents/custom-element` polyfill is calling `attributeChangedCallback()` twice
- Safari Tech Preview's `customElements.get()` name parameter is case sensitive

This PR is for moving to the new polyfill and fixing the issues associated with the move.

### PR DESCRIPTION
**Pull request by treshugart**
fixes #908

### CODE REVIEW

**Comment by treshugart** [comment]
It did because of https://github.com/w3c/webcomponents/issues/604.

File: src/api/vdom.js [hunk 1]
```diff
@@ -89,7 +89,7 @@ const attributesContext = propContext(attributes, {
 
   // Default attribute applicator.
   [symbols.default] (elem, name, value) {
-    const { props, prototype } = customElements.get(elem.tagName) || {
+    const { props, prototype } = customElements.get(elem.tagName.toLowerCase()) || {
```
Comment (treshugart) [inline src/api/vdom.js]: @bengummer this is what Safari TP would fail on before.

File: src/api/component.js [hunk 2]
```diff
@@ -19,6 +20,22 @@ import root from 'window-or-global';
...
 
+
+
+// Prevent double-calling with polyfill.
+
+const $prevName = createSymbol();
+const $prevOldValue = createSymbol();
+const $prevNewValue = createSymbol();
+
+function preventDoubleCalling (elem, name, oldValue, newValue) {
```
Comment (treshugart) [inline src/api/component.js]: This fixes the polyfill.

File: src/api/component.js [hunk 3]
```diff
@@ -237,6 +254,16 @@ Component.prototype = Object.create(htmlElementPrototype, {
   attributeChangedCallback: {
     configurable: true,
     value (name, oldValue, newValue) {
+      // Polyfill calls this twice.
+      if (preventDoubleCalling(this, name, oldValue, newValue)) {
+        return;
+      }
+
+      // Set data so we can prevent double calling if the polyfill.
+      this[$prevName] = name;
```
Comment (treshugart) [inline src/api/component.js]: The other part of fixing the polyfill.

File: src/api/vdom.js [hunk 4]
```diff
@@ -89,7 +89,7 @@ const attributesContext = propContext(attributes, {
 
   // Default attribute applicator.
   [symbols.default] (elem, name, value) {
-    const { props, prototype } = customElements.get(elem.tagName) || {
+    const { props, prototype } = customElements.get(elem.localName) || {
```
Comment (treshugart) [inline src/api/vdom.js]: Fun fact, seems `localName` returns lowercase. We should be using it anyways, but it's a bonus it's lowercase.

File: package.json [hunk 5]
```diff
@@ -32,7 +32,9 @@
     "react": "15.x",
     "react-dom": "15.x",
     "skatejs-build": "12.x",
-    "skatejs-web-components": "3.x"
+    "@webcomponents/custom-elements": "https://github.com/webcomponents/custom-elements/archive/fc022e8.tar.gz",
+    "@webcomponents/shadycss": "https://github.com/webcomponents/shadycss/archive/19a46c9.tar.gz",
+    "@webcomponents/shadydom": "https://github.com/webcomponents/shadydom/archive/c162774.tar.gz"
```
Comment (jpnelson) [inline package.json]: would we keep skatejs/web-components up to date with these?

File: src/api/component.js [hunk 6]
```diff
@@ -19,6 +20,22 @@ import root from 'window-or-global';
...
+
+
+// Prevent double-calling with polyfill.
+
+const $prevName = createSymbol();
+const $prevOldValue = createSymbol();
+const $prevNewValue = createSymbol();
+
+function preventDoubleCalling (elem, name, oldValue, newValue) {
+  return name === elem[$prevName]
```
Comment (justinfagnani) [inline src/api/component.js]: Can you send me a repro of the original issue?

Comment (treshugart) [inline src/api/component.js]: I've created a basic reproduction of what we're doing internally: http://www.webpackbin.com/EyVC4-Agz.

I think we should just continue being defensive, but this does raise some questions. In that bin:

- Chrome CE v1 text content is 1
- Safari CE v1 text content is 1772 (or whatever, call stack exceeded)
- In polyfill text content is 2

File: src/api/vdom.js (hunk 4 above)
Comment (jpnelson) [inline src/api/vdom.js]: should we be normalising it here then? Or is `customElements.get` not case sensitive?

Comment (treshugart) [inline src/api/vdom.js]: It's case-sensitive, but `localName` should return the exact name of the element how it was specified in HTML, as I understand. More info here https://github.com/w3c/webcomponents/issues/604.

**Comment by treshugart** [comment]
I'll update the skatejs-web-components stuff later. That PR is taking a bit longer because we need o figure out a way to consume the native-shim in both native and non-native browsers. Conditional requiring is difficult.

**Pull request merged**